test/storage/recovery_test.cpp

```diff
@@ -0,0 +1,533 @@
+
+  // Create the table
+  txn = txn_manager.BeginTransaction();
+  auto db_catalog = catalog.GetDatabaseCatalog(txn, db_oid);
+  auto table_schema = StorageTestUtil::RandomSchemaNoVarlen(5, &generator_);
+  auto table_oid = db_catalog->CreateTable(txn, namespace_oid, table_name, *table_schema);
+  EXPECT_TRUE(table_oid != catalog::INVALID_TABLE_OID);
+  auto *table_ptr = new storage::SqlTable(&block_store_, *table_schema);
+  EXPECT_TRUE(db_catalog->SetTablePointer(txn, table_oid, table_ptr));
+  txn_manager.Commit(txn, transaction::TransactionUtil::EmptyCallback, nullptr);
+
+  // Create an index
+  txn = txn_manager.BeginTransaction();
+  auto index_schema = DummyIndexSchema();
+  auto index_oid = db_catalog->CreateIndex(txn, namespace_oid, index_name, table_oid, index_schema);
+  EXPECT_TRUE(index_oid != catalog::INVALID_INDEX_OID);
+  auto *index_ptr = storage::index::IndexBuilder()
+                        .SetConstraintType(storage::index::ConstraintType::UNIQUE)
+                        .SetKeySchema(index_schema)
+                        .SetOid(index_oid)
+                        .Build();
+  EXPECT_TRUE(db_catalog->SetIndexPointer(txn, index_oid, index_ptr));
+  txn_manager.Commit(txn, transaction::TransactionUtil::EmptyCallback, nullptr);
+
+  // Drop the database
+  txn = txn_manager.BeginTransaction();
+  EXPECT_TRUE(catalog.DeleteDatabase(txn, db_oid));
+  txn_manager.Commit(txn, transaction::TransactionUtil::EmptyCallback, nullptr);
+  EXPECT_EQ(catalog::INVALID_DATABASE_OID, catalog.GetDatabaseOid(txn, database_name));
+  EXPECT_FALSE(catalog.GetDatabaseCatalog(txn, db_oid));
+
+  // Guarantee persist of log records
+  log_manager.PersistAndStop();
+
+  /* CRASH */
+
+  // Start a transaction manager with logging disabled, we don't want to log the log replaying
+  transaction::TransactionManager recovery_txn_manager{&pool_, true, LOGGING_DISABLED};
+
+  // Create catalog for recovery
+  catalog::Catalog recovered_catalog(&recovery_txn_manager, &block_store_);
+
+  // Instantiate recovery manager, and recover the catalog.
+  DiskLogProvider log_provider(LOG_FILE_NAME);
+  RecoveryManager recovery_manager(&log_provider, common::ManagedPointer(&recovered_catalog), &recovery_txn_manager,
+                                   common::ManagedPointer(&thread_registry_), &block_store_);
+  recovery_manager.StartRecovery();
+  recovery_manager.FinishRecovery();
+
+  // Assert the database does not exist
+  txn = recovery_txn_manager.BeginTransaction();
+  EXPECT_EQ(catalog::INVALID_DATABASE_OID, recovered_catalog.GetDatabaseOid(txn, database_name));
+  EXPECT_FALSE(recovered_catalog.GetDatabaseCatalog(txn, db_oid));
+  recovery_txn_manager.Commit(txn, transaction::TransactionUtil::EmptyCallback, nullptr);
+
+  log_manager.Start();
+  GarbageCollectCatalogObject(&catalog, &txn_manager);
+  log_manager.PersistAndStop();
+  GarbageCollectCatalogObject(&recovered_catalog, &recovery_txn_manager);
+  delete table_schema;
+}
```

Comment:
Copying from another comment since it more explicitly states what I tried to describe in the third bullet above.

```
T1               T2
                 BEGIN
BEGIN
DROP table1
COMMIT

[GC runs several times]

                 INSERT a INTO table1
                 INSERT b INTO table2
                 COMMIT
```